### PR TITLE
Add support for reading from stdin

### PR DESCRIPTION
**Pull request by thisch**
First of all thx @cpitclaudel for starting with the first draft (#1189) of the Read-from-stdin feature.

This PR is based on the inital PR by @cpitclaudel and tries to incorporate the comment by @PCManticore  from #1189. I would like to know from @PCManticore if this PR now goes into the right direction.

I tested this PR by using the following command:
```
cat examples/custom_raw.py | ./bin/pylint --from-stdin custom_raw.py 
```

Closes: #1187

## Steps

- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
| ✓  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
| ✓  | :scroll: Docs |

### CODE REVIEW

**Comment by PCManticore** [comment]
@thisch This is very cool, thanks for working on it! Can you add some tests as well?

File: pylint/lint.py [hunk 1]
```diff
@@ -89,6 +90,24 @@
...
+        return MANAGER.astroid_cache[modname]
+    from astroid.builder import AstroidBuilder
+
+    return AstroidBuilder(MANAGER).string_build(data, modname, filepath)
+
+
+def _read_stdin():
+    # https://mail.python.org/pipermail/python-list/2012-November/634424.html
+    # FIXME should this try to check the file's declared encoding?
+    sys.stdin = TextIOWrapper(sys.stdin.detach(), newline=None)
```
Comment (thisch) [inline pylint/lint.py]: I don't yet understand why we need this line.

Comment (thisch) [reply pylint/lint.py]: This probably helps: https://stackoverflow.com/questions/31488419/binary-input-with-an-ascii-text-header-read-from-stdin

**Comment by thisch** [comment]
I need a bit your help with fixing the unit tests:

Here are the errors reported by travis ci:
```
3.4:
AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in
                       '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
3.5-dev:
	
AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
                       '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'

3.6:
AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
                       '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'

3.7:
AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
                       '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
```

Where does this "M...0611" output come from? When I run pylint on a file with the contents "import os" locally, pylint reports

```
************* Module mymodule
mymodule.py:1:0: C0111: Missing module docstring (missing-docstring)
mymodule.py:1:0: W0611: Unused import os (unused-import)

----------------------------------------------------------------------
Your code has been rated at -10.00/10 (previous run: -10.00/10, +0.00)
```
I don't understand why the travis-ci output contains "M...0611" whereas locally I get "W0611".

File: CONTRIBUTORS.txt [hunk 2]
```diff
@@ -269,3 +269,5 @@ contributors:
 * Svetoslav Neykov: contributor
 
 * Federico Bond: contributor
+
+* Thomas Hisch: contributor
```
Comment (cpitclaudel) [inline CONTRIBUTORS.txt]: If you patch is based on the one in #1189, please acknowledge that here too :)

**Comment by cpitclaudel** [comment]
Thanks for working on this! I don't have time to review the patch, but I'm excited to see that this feature will be coming soon ^^

File: pylint/lint.py [hunk 3]
```diff
@@ -89,6 +90,24 @@
 MANAGER = astroid.MANAGER
 
 
+def _ast_from_string(data, filepath, modname):
+    if (
```
Comment (PCManticore) [inline pylint/lint.py]: Rather than doing three lookups, let's just do one:
```suggestion
    cached = MANAGER.astroid_cache.get(modname)
    if cached and cached.file == filepath:
         return cached
```

File: pylint/lint.py [hunk 4]
```diff
@@ -1054,31 +1083,51 @@ def _do_check(self, files_or_modules):
             if interfaces.implements(checker, interfaces.IAstroidChecker):
                 walker.add_checker(checker)
         # build ast and check modules or packages
-        for descr in self.expand_files(files_or_modules):
-            modname, filepath, is_arg = descr["name"], descr["path"], descr["isarg"]
-            if not self.should_analyze_file(modname, filepath, is_argument=is_arg):
-                continue
+        if self.config.from_stdin:
+            assert len(files_or_modules) == 1
```
Comment (PCManticore) [inline pylint/lint.py]: I'm not sure I understand why we need a file here? Can't we output the report to the stdout? If we need a filename for reporting purposes, wouldn't be enough to provide a fake one, e.g. `<string>` or something of sorts?

File: pylint/test/test_self.py [hunk 5]
```diff
@@ -113,6 +114,7 @@ def _runtest(self, args, reporter=None, out=None, code=None):
 
     def _run_pylint(self, args, out, reporter=None):
         args = args + ["--persistent=no"]
+        sys.argv = ["pylint"] + args
```
Comment (PCManticore) [inline pylint/test/test_self.py]: Why do we add `pylint` here?

File: CONTRIBUTORS.txt (hunk 2 above)
Comment (thisch) [reply CONTRIBUTORS.txt]: Sure. Thx for pointing that out.

File: pylint/test/test_self.py (hunk 5 above)
Comment (thisch) [reply pylint/test/test_self.py]: Because sys.argv[0] is used in the Usage help text, which is checked by one of my tests.
Instead of setting sys.argv here, we could set it only for the test(s) that need it or we simply only check the return value of the pylint call (via the code kwargs of _runtests). I would prefer the latter.

File: pylint/lint.py (hunk 4 above)
Comment (thisch) [reply pylint/lint.py]: I've seen that providing a fake filename is also supported by other linters
See e.g. https://github.com/jshint/jshint/pull/1503 and https://eslint.org/docs/user-guide/command-line-interface

Comment (cpitclaudel) [reply pylint/lint.py]: I think this matters in particular for (relative) imports.

**Comment by thisch** [comment]
> I need a bit your help with fixing the unit tests:
> 
> Here are the errors reported by travis ci:
> 
> ```
> 3.4:
> AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in
>                        '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
> 3.5-dev:
> 	
> AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
>                        '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
> 
> 3.6:
> AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
>                        '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
> 
> 3.7:
> AssertionError: assert '************* Module mymodule\nmymodule.py:1:0: W0611: Unused import os (unused-import)' in 
>                        '************* Module mymodule\nmymodule.py:1:0: C0304: Final newline missing (missing-final-newline)\n************* M...0611: Unused import os (unused-import)\n\n-------------------------------------\nYour code has been rated at -20.00/10'
> ```
> 
> Where does this "M...0611" output come from? When I run pylint on a file with the contents "import os" locally, pylint reports
> 
> ```
> ************* Module mymodule
> mymodule.py:1:0: C0111: Missing module docstring (missing-docstring)
> mymodule.py:1:0: W0611: Unused import os (unused-import)
> 
> ----------------------------------------------------------------------
> Your code has been rated at -10.00/10 (previous run: -10.00/10, +0.00)
> ```
> 
> I don't understand why the travis-ci output contains "M...0611" whereas locally I get "W0611".

The "M...0611" output is generated by the pytest, so it is not related to an issue in pylint.

File: pylint/lint.py (hunk 4 above)
Comment (thisch) [reply pylint/lint.py]: You are right, handling of relative imports needs to be taken into account. I'll write a testcase for that and then we can discuss how to fix the issue.

Comment (thisch) [reply pylint/lint.py]: here is the testcase https://github.com/thisch/pylint/commit/5583543d9c9e780ea84dde0697e38b7ccafea24a

the results output by the `--from-stdin` run differ from the normal (non-stdin) run.

Can anybody explain why the relative imports are not working in the `--from-stdin` case?

File: pylint/test/test_self.py (hunk 5 above)
Comment (PCManticore) [reply pylint/test/test_self.py]: Yeah, the latter sounds better for me as well.

File: pylint/lint.py (hunk 4 above)
Comment (thisch) [reply pylint/lint.py]: I've just fixed a bug in the code that determines the module name of the file passed as a positional argument to pylint (after --from-stdin).

To determine the correct modulename I copied the code from pylint.utils.expand_modules. I hope that I didn't miss anything.

The ``--from-stdin`` feature should now work as expected.

**Comment by thisch** [comment]
@PCManticore when you have time, please take again a look at this PR. I think that it is ready now.

File: pylint/lint.py [hunk 6]
```diff
@@ -89,6 +90,23 @@
 MANAGER = astroid.MANAGER
 
 
+def _ast_from_string(data, filepath, modname):
+    cached = MANAGER.astroid_cache.get(modname)
+    if cached and cached.file == filepath:
+        return cached
+
+    from astroid.builder import AstroidBuilder
```
Comment (PCManticore) [inline pylint/lint.py]: We can import the builder at the top btw, there's no circular import to be avoided here.

File: pylint/lint.py [hunk 7]
```diff
@@ -1054,31 +1082,57 @@ def _do_check(self, files_or_modules):
...
-        for descr in self.expand_files(files_or_modules):
-            modname, filepath, is_arg = descr["name"], descr["path"], descr["isarg"]
-            if not self.should_analyze_file(modname, filepath, is_argument=is_arg):
-                continue
+        if self.config.from_stdin:
+            assert len(files_or_modules) == 1
+            filepath = files_or_modules[0]
+            try:
+                # Note that modpath_from_file does not really perform an
+                # __import__ as one may expect, because ImportError
```
Comment (PCManticore) [inline pylint/lint.py]: The comment seems a bit confusing, indeed the function doesn't perform an `__import__`, but it raises an `ImportError`, which is why we should handle it here. Maybe rephrase the comment a bit?

File: pylint/test/test_self.py [hunk 8]
```diff
@@ -546,3 +547,79 @@ def test_parseable_file_path(self):
...
+
+            # this code needs to work w/ and w/o a file named a/b.py on the
+            # harddisk.
+            with mock.patch("pylint.lint._read_stdin", return_value=b_code):
+                self._test_output(
+                    ["--from-stdin", join("a", "b.py")], expected_output=expected
+                )
+
+        finally:
+            os.chdir(curdir)
```
Comment (PCManticore) [inline pylint/test/test_self.py]: Great test, thanks!

File: pylint/lint.py [hunk 9]
```diff
@@ -1054,31 +1082,57 @@ def _do_check(self, files_or_modules):
             if interfaces.implements(checker, interfaces.IAstroidChecker):
                 walker.add_checker(checker)
         # build ast and check modules or packages
-        for descr in self.expand_files(files_or_modules):
-            modname, filepath, is_arg = descr["name"], descr["path"], descr["isarg"]
-            if not self.should_analyze_file(modname, filepath, is_argument=is_arg):
-                continue
+        if self.config.from_stdin:
+            assert len(files_or_modules) == 1
```
Comment (PCManticore) [inline pylint/lint.py]: Maybe raise a more obvious exception? if we don't have anything, you can add one in `pylint.exceptions`.

File: pylint/lint.py [hunk 10]
```diff
@@ -1054,31 +1082,57 @@ def _do_check(self, files_or_modules):
...
-                self.msgs_store
-            )
-            for msgid, line, args in spurious_messages:
-                self.add_message(msgid, line, None, args)
+            ast_node = _ast_from_string(_read_stdin(), filepath, modname)
+
+            if ast_node is not None:
+                self.file_state = utils.FileState(filepath)
+                self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
+                # warn about spurious inline messages handling
```
Comment (PCManticore) [inline pylint/lint.py]: This block seems duplicated with the normal flow, maybe move it after the whole `if` branch?

Comment (thisch) [reply pylint/lint.py]: Moving it after whole `if self.config.from_stdin` branch does not work IMO. I think it is not worth the effort to avoid this code duplication, especially since only a couple of lines of code are duplicated.

**Comment by thisch** [comment]
the failure on travis-ci is unrelated to my change.

**Comment by PCManticore** [comment]
Thank you @thisch for your contribution! 🎉 This will be part of pylint 2.4. when it gets released.

**Comment by thisch** [comment]
@PCManticore  Any estimates when we can expect a 2.4 release? What are the open issues that need to be fixed before 2.4 can be relased?

**Comment by PCManticore** [comment]
@thisch It's going to happen most likely in July. A main blocker is having Python 3.8 support. A separate one is that I'd like to have https://github.com/PyCQA/pylint/pull/2036 in time for 2.4 release, but if that won't happen in due time, we can release without it.
